Mathlib/Data/Multiset/Antidiagonal.lean

```diff
@@ -0,0 +1,127 @@
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Multiset.Powerset
+
+/-!
+# The antidiagonal on a multiset.
+
+The antidiagonal of a multiset `s` consists of all pairs `(t₁, t₂)`
+such that `t₁ + t₂ = s`. These pairs are counted with multiplicities.
+-/
+
+
+namespace Multiset
+
+open List
+
+variable {α β : Type _}
+
+/-- The antidiagonal of a multiset `s` consists of all pairs `(t₁, t₂)`
+    such that `t₁ + t₂ = s`. These pairs are counted with multiplicities. -/
+def antidiagonal (s : Multiset α) : Multiset (Multiset α × Multiset α) :=
+  Quot.liftOn s (fun l ↦ (revzip (powersetAux l) : Multiset (Multiset α × Multiset α)))
+    fun _ _ h ↦ Quot.sound (revzip_powersetAux_perm h)
+#align multiset.antidiagonal Multiset.antidiagonal
+
+theorem antidiagonal_coe (l : List α) : @antidiagonal α l = revzip (powersetAux l) :=
+  rfl
+#align multiset.antidiagonal_coe Multiset.antidiagonal_coe
+
+@[simp]
+theorem antidiagonal_coe' (l : List α) : @antidiagonal α l = revzip (powersetAux' l) :=
+  Quot.sound revzip_powersetAux_perm_aux'
+#align multiset.antidiagonal_coe' Multiset.antidiagonal_coe'
+
+/- Porting note: Mathport seems to be replacing `quotient.induction_on s $` with
+`(Quotient.induction_on s)`, when it should be `Quotient.inductionOn s <|`. I'm more concerned
+about the parentheses being used instead of `<|` than the typo. -/
+/- Porting note: `simp` seemed to be applying `antidiagonal_coe'` instead of `antidiagonal_coe`
+in what used to be `simp [antidiagonal_coe]`. -/
+/-- A pair `(t₁, t₂)` of multisets is contained in `antidiagonal s`
+    if and only if `t₁ + t₂ = s`. -/
+@[simp]
+theorem mem_antidiagonal {s : Multiset α} {x : Multiset α × Multiset α} :
+    x ∈ antidiagonal s ↔ x.1 + x.2 = s :=
+  Quotient.inductionOn s <| fun l ↦ by
+    dsimp only [quot_mk_to_coe, antidiagonal_coe]
+    refine' ⟨fun h => revzip_powersetAux h, fun h ↦ _⟩
+    haveI := Classical.decEq α
+    simp only [revzip_powersetAux_lemma l revzip_powersetAux, h.symm, ge_iff_le, mem_coe, mem_map',
+  mem_powersetAux]
+    cases' x with x₁ x₂
+    exact ⟨x₁, le_add_right _ _, by rw [add_tsub_cancel_left x₁ x₂]⟩
+#align multiset.mem_antidiagonal Multiset.mem_antidiagonal
+
+@[simp]
+theorem antidiagonal_map_fst (s : Multiset α) : (antidiagonal s).map Prod.fst = powerset s :=
+  Quotient.inductionOn s <| fun l ↦ by simp [powersetAux'];
+#align multiset.antidiagonal_map_fst Multiset.antidiagonal_map_fst
+
+@[simp]
+theorem antidiagonal_map_snd (s : Multiset α) : (antidiagonal s).map Prod.snd = powerset s :=
+  Quotient.inductionOn s <| fun l ↦ by simp [powersetAux']
+#align multiset.antidiagonal_map_snd Multiset.antidiagonal_map_snd
+
+/- Porting note: I changed `@antidiagonal` to `@antidiagonal.{u_1}` because otherwise I got
+`Multiset.antidiagonal_zero.{u_2, u_1} {α : Type (max u_1 u_2)} : ...`, which gave me issues
+and triggered the linter. -/
+@[simp]
+theorem antidiagonal_zero : @antidiagonal.{u_1} α 0 = {(0, 0)} :=
+  rfl
```

Comment:
This is weird. I don't know what's going on, but it would be good to understand this better, I think.

cc @digama0